src/Access.ts

```diff
@@ -1,25 +1,33 @@
 import { AxiosInstance } from 'axios';
 import { BigNumber, ethers } from 'ethers';
 import { readHash } from './constant';
-import { makeTx, parseHex, Arcana, customError, ensureArray } from './Utils';
+import { makeTx, parseHex, Arcana, customError, ensureArray, getAppAddress } from './Utils';
 
 export class Access {
   private provider: any;
   private api: AxiosInstance;
   private appAddress: string;
 
-  constructor(appAddress: string, provider: any, api: AxiosInstance) {
+  constructor(appAddress: string | undefined, provider: any, api: AxiosInstance) {
     this.provider = provider;
     this.api = api;
     this.appAddress = appAddress;
   }
 
-  share = async (fileId: string[] | string, _address: string[] | string, validity: number[] | number | null): Promise<string> => {
-    fileId = ensureArray(fileId).map(parseHex)
-    const address = ensureArray(_address).map(parseHex)
+  setAppAddress = async (did: string) => {
+    // if app address is not set then fetch it from the did and set it
+    if (!this.appAddress) {
+      this.appAddress = await getAppAddress(parseHex(did), this.provider);
+    }
+  }
 
+
+  share = async (did: string[] | string, _address: string[] | string, validity: number[] | number | null): Promise<string> => {
+    did = ensureArray(did).map(parseHex)
+    await this.setAppAddress(did[0]);
+    const address = ensureArray(_address).map(parseHex)
     const accessType = [];
-    fileId.forEach(() => {
+    did.forEach(() => {
```

Comment:
Should it be synchronous (for of ) than asynchronous(forEach) ?

Reply:
Not required, I guess